README.md
# JSON API Mapper

Implemtation of [JSON API Standard Specification](https://jsonapi.org/)

This project goal is to create easy-to-use library to implement JSON API specification.

Whole project is at the beginning of development. So don't hesitate to help. I'm open to some good ideas how make this
more customizable and friendly.

Library only provides wrappers to create valid JSON API document. Controllers and Response is on you.

## Issues

You can write [email](mailto://incoming+bednic-json-api-10827057-issue-@incoming.gitlab.com) or create issue
in [gitlab](https://gitlab.com/bednic/json-api/issues)

## Installation

Install library via Composer

```
composer require bednic/json-api
```

## Basic Usage

> For simplicity, we use `$container` as some dependency provider (Dependency Injection).

### MetadataRepository

First we need create `MetadataRepository`. That we get from `MetadataFactory`.

```php
<?php
/** @var $container Psr\Container\ContainerInterface */
// This is cache instance implements PSR SimpleCache
$cache = $container->get( Psr\SimpleCache\CacheInterface::class);
// This is AnnotationDriver or SchemaDriver, depends on your preferences
$driver = $container->get(\JSONAPI\Driver\Driver::class);
// Paths to your object representing resources
$paths = ['paths/to/your/resources','another/path'];
// Factory returns instance of MetadataRepository
$repository = JSONAPI\Factory\MetadataFactory::create(
            $paths,
            $cache,
            $driver
        );

```

### DocumentBuilderFactory

#### Options

|Param|Default|Description|
|-----|-------|-----------|
|metadataRepository| |Instance of MetadataRepository.|
|baseURL| |URL where you API lays.|
|maxIncludedItems|625|Maximum items included in Compound Document. This prevents to compound huge documents.|
|relationshipLimit|25|Maximum items included in relationship collection.|
|relationshipData|true|Toggle if relationships should have data attribute.|
|supportInclusion|true|Toggle if server supports [inclusion](https://jsonapi.org/format/#fetching-includes).|
|supportSort|true|Toggle if server supports [sort](https://jsonapi.org/format/#fetching-sorting).|
|supportPagination|true|Toggle if server supports [pagination](https://jsonapi.org/format/#fetching-pagination).|
|paginationParser|null|PaginationParserInterface instance, which is responsible for parsing pagination.|
|filterParser|null|FilterParserInterface instance, which is responsible for parsing filter.|
|logger|null|LoggerInterface instance, PSR compliant logger instance.|

```php
<?php

// First we need DocumentBuilderFactory
// Let's get MetadataRepository from DI
/** @var $container Psr\Container\ContainerInterface */
$metadataRepository = $container->get(JSONAPI\Metadata\MetadataRepository::class);
$baseURL = "http://localhost/"; // base URL where API lays
// You can initialize it by yourself
$dbf = new JSONAPI\Factory\DocumentBuilderFactory($metadataRepository, $baseURL);
// OR you can let DI make it for you
$dbf = $container->get(JSONAPI\Factory\DocumentBuilderFactory::class);
// We need PSR ServerRequestInterface instance
/** @var  $request Psr\Http\Message\ServerRequestInterface */
// Here we create instance of JSONAPI\Document\Builder
$builder = $dbf->new($request);
// Your object|objects
$data = new MyObject();
// $doc here is Document instance
$doc = $builder->setData($data)->build();

```

Now if you encode `$doc` to JSON, you're going to get full JSON API Document. More information about UriParser
is [here](UriParser)

### Describing your objects

You can choose which way you want to describe your object metadata.

#### With Annotations

> Note: If you want to use annotations you have to use `AnnotationDriver` in `MetadataFactory`

### [Example](tests-resources/valid/GettersExample.php)

How you can see, setting up resource object is quiet easy. Just annotate your getter with ``#[Attribute]``
or ``#[Relationship]`` annotation.

#### Schema

The important part is to implement Resource interface. Then fill up static method `getSchema`.

> Note: If you want to use schema you have to use `SchemaDriver` in `MetadataFactory`

### [Example](tests-resources/valid/PropsExample.php)

## Open API Schema

This library provides lightweight wrapper around OAS. It can generate OAS v3.0.3 schema in json,
so you can provide doc for your api easily.

### Basic Example

```php
    $factory = new OpenAPISpecificationBuilder(
        $metadataRepository,
        'https://your.api.url'
    );

    $info = new Info('JSON:API OAS', '1.0.0');
    $info->setDescription('Test specification');
    $info->setContact(
        (new Contact())
            ->setName('Tomas Benedikt')
            ->setEmail('tomas.benedikt@gmail.com')
            ->setUrl('https://gitlab.com/bednic')
    );
    $info->setLicense(
        (new License('MIT'))
            ->setUrl('https://gitlab.com/bednic/json-api/-/blob/5.x/LICENSE')
    );
    $info->setTermsOfService('https://gitlab.com/bednic/json-api/-/blob/5.x/CONTRIBUTING.md');

    $oas = $factory->create($info);
    $oas->setExternalDocs(new ExternalDocumentation('https://gitlab.com/bednic/json-api/-/wikis/home'));

    $json = json_encode($oas);
```

## UriParser

This object works with url, and parse required keywords as described at JSON API Standard

### Options

|Param|Default|Description|
|-----|-------|-----------|
|request| |Instance of PSR compliant ServerRequestInterface.|
|metadataRepository| |Instance of MetadataRepository.|
|baseURL| |URL where you API lays.|
|supportInclusion|true|Toggle if server supports [inclusion](https://jsonapi.org/format/#fetching-includes).|
|supportSort|true|Toggle if server supports [sort](https://jsonapi.org/format/#fetching-sorting).|
|supportPagination|true|Toggle if server supports [pagination](https://jsonapi.org/format/#fetching-pagination).|
|paginationParser|null|PaginationParserInterface instance, which is responsible for parsing pagination.|
|filterParser|null|FilterParserInterface instance, which is responsible for parsing filter.|
|logger|null|LoggerInterface instance, PSR compliant logger instance.|

### PathInterface

Provides information about path, like resource type, resource ID, relation type, is it collection or is it relationship

### Fieldset

https://jsonapi.org/format/#fetching-sparse-fieldsets

### Filter

https://jsonapi.org/format/#fetching-filtering

As described, specification is agnostic about filter implementation. So I created, more like borrowed, expression filter
from OData. So now you can use something like this:

`filter=stringProperty eq 'string' and contains(stringProperty,'asdf') and intProperty in (1,2,3) or boolProperty ne true and relation.property eq null`

There are for now two `ExpressionBuilder`s:

* `DoctrineQueryExpressionBuilder` creates expression for Doctrine QueryBuilder, it comes hadny if you are using
  Doctrine and want more complex filtering
* `DoctrineCriteriaExpressionBuilder` works well with ArrayCollection or PersistentCollection which is dependency of
  this library so it's default and you can use it too. But it provides only some basic expressions like eq, ne, etc...
  But you can't filter by relation property or use some functions, or expressions.
* `ClosureExpressionBuilder` create expression tree from closures, so you do not need doctrine at all.

### Pagination

https://jsonapi.org/format/#fetching-pagination

I implement two of three pagination technics

* LimitOffsetPagination
* PagePagination
* CursorPagination which is only abstract, cause it need more then just several number, so if you want use cursor based
  pagination, you have to implement it by yourself.

### Includes

https://jsonapi.org/format/#fetching-includes

### Sort

https://jsonapi.org/format/#fetching-sorting

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [10.1.0](https://gitlab.com/bednic/json-api/compare/10.0.3...10.1.0) (2022-06-14)


### Added

* add has operator ([a986931](https://gitlab.com/bednic/json-api/commit/a9869311e9098f612570e44173974a250196fe47))

### [10.0.3](https://gitlab.com/bednic/json-api/compare/10.0.2...10.0.3) (2022-05-24)


### Fixed

* fix Driver read array data type if docblock does not exist ([0039ac8](https://gitlab.com/bednic/json-api/commit/0039ac8048fe6825d11721011b464b6963e6919a))
* fix Driver read array data type if docblock exist on property ([173b26a](https://gitlab.com/bednic/json-api/commit/173b26a80ac3a6931de489e07ce9a2c86d8f3973))
* PathParser did not assign relationship name if related path ([432dccc](https://gitlab.com/bednic/json-api/commit/432dcccb2c2c62d88c9cece1988dabf8df9b62d1))

### [10.0.2](https://gitlab.com/bednic/json-api/compare/10.0.1...10.0.2) (2022-05-23)

### [10.0.1](https://gitlab.com/bednic/json-api/compare/10.0.0...10.0.1) (2022-05-23)

## [10.0.0](https://gitlab.com/bednic/json-api/compare/9.0.6...10.0.0) (2022-05-22)


### ⚠ BREAKING CHANGES

* FilterParserInterface::parse method signature changed
* Filtering/Builder/* removed
ExpressionFilterParser ctr
changed

### Changed

* allow all chars except '/' for identifier ([3c18f0f](https://gitlab.com/bednic/json-api/commit/3c18f0fec1e92d3daa3fa5201fdccae447159492))
* allow more special characters in ID ([b0e69a7](https://gitlab.com/bednic/json-api/commit/b0e69a7076cc0bf30870d664a9fe0312d4f82271))
* Change passing PathInterface to FilterParser ([355154f](https://gitlab.com/bednic/json-api/commit/355154f46826261696d1cce7a0f25cc2bd946ab5))
* QuatrodotResult ([e27a28b](https://gitlab.com/bednic/json-api/commit/e27a28bd4c96dd36f2eb750a445da44086a2804c))
* Remove builders ([fd2f13c](https://gitlab.com/bednic/json-api/commit/fd2f13c7b6f919c97422ebe0d71526abe8813818))
* specify return types ([3b8633b](https://gitlab.com/bednic/json-api/commit/3b8633b9766356d712a649a3455821df0781132d))


### Fixed

* fix issue 44 ([774fe6d](https://gitlab.com/bednic/json-api/commit/774fe6d962e4bc8193b940e3a27e60dd8f63e6d1))
* fix pattern match ([88eedba](https://gitlab.com/bednic/json-api/commit/88eedbaae6d9c30e645c39c26c28f83220517c80))
* null identifierExpressions handling ([aaca24a](https://gitlab.com/bednic/json-api/commit/aaca24a06b1d06e49b1d9043236d9cbbe9157e49))
* null identifierExpressions handling ([509f689](https://gitlab.com/bednic/json-api/commit/509f68996d32e21542a87d85fbe0606b97964180))
* safe url encode URI parts ([c97e5a4](https://gitlab.com/bednic/json-api/commit/c97e5a4d3722eabcc9f8858dbb270dc51e366efd)), closes [#45](https://gitlab.com/bednic/json-api/issues/45)
* safe url encode URI parts ([0f6a045](https://gitlab.com/bednic/json-api/commit/0f6a045336833f92175d2c22ba2160708da977a1)), closes [#45](https://gitlab.com/bednic/json-api/issues/45)

### [9.0.8](https://gitlab.com/bednic/json-api/compare/9.0.7...9.0.8) (2022-04-19)


### Fixed

* fix pattern match ([85380d9](https://gitlab.com/bednic/json-api/commit/85380d9459432cf6684f3e3636ffa858c552494b))

### [9.0.7](https://gitlab.com/bednic/json-api/compare/9.0.6...9.0.7) (2022-04-19)


### Fixed

 * fix issue 44 ([774fe6d9](https://gitlab.com/bednic/json-api/-/commit/774fe6d962e4bc8193b940e3a27e60dd8f63e6d1))

### [9.0.6](https://gitlab.com/bednic/json-api/compare/9.0.5...9.0.6) (2022-02-13)


### Changed

* Stan to lvl 6, Remove Symfony ClassMapLoader, Remove MyClabs Enum, PHP 8.1 ([73b7a15](https://gitlab.com/bednic/json-api/commit/73b7a153469a2f4c01d4857c419fde581864094f))


### Fixed

* Stan error ([4053bc9](https://gitlab.com/bednic/json-api/commit/4053bc9b22ac6bbac4e23b9faff43cd225655fc3))

### [9.0.5](https://gitlab.com/bednic/json-api/compare/9.0.4...9.0.5) (2022-02-07)

### [9.0.4](https://gitlab.com/bednic/json-api/compare/9.0.3...9.0.4) (2022-02-07)

### [9.0.3](https://gitlab.com/bednic/json-api/compare/9.0.2...9.0.3) (2022-02-07)


### Fixed

* URI parsing sequence ([e41a3fa](https://gitlab.com/bednic/json-api/commit/e41a3fac86d87c2cb87d3f05109d9ed366d348f5))

### [9.0.2](https://gitlab.com/bednic/json-api/compare/9.0.1...9.0.2) (2021-11-24)


### Fixed

* clear relationship and isRelationship properties in PathParser::parse() ([10dfe7d](https://gitlab.com/bednic/json-api/commit/10dfe7d8d2fafaacbfe2ef80ce4b860231940a7e)), closes [#88](https://gitlab.com/bednic/json-api/issues/88)

### [9.0.1](https://gitlab.com/bednic/json-api/compare/9.0.0...9.0.1) (2021-11-24)


### Fixed

* get size returns emtpy even it it is not empty ([02599b6](https://gitlab.com/bednic/json-api/commit/02599b6818aef55dd23cb1137cd36ec572249acd))
* remove usage of body::getSize ([3adb78d](https://gitlab.com/bednic/json-api/commit/3adb78d06f3016a10482ca0f3be530e6b3d4efd3))
* usage of content in middleware ([a785a9f](https://gitlab.com/bednic/json-api/commit/a785a9faeca254eb0a03315a74843f72c8f00d19))

## [9.0.0](https://gitlab.com/bednic/json-api/compare/8.1.0...9.0.0) (2021-11-10)


### ⚠ BREAKING CHANGES

* * Factory namespace removed
* Configuration class instead of parameters
* ParsedURI interface instead URIParser

### Added

* Add Builder instance and URIParser instance to ServerRequestInstance ([908b2a4](https://gitlab.com/bednic/json-api/commit/908b2a4bb24118a61dc1c524550e91661e3e1c58))
* Introduce Configuration ([d26b1df](https://gitlab.com/bednic/json-api/commit/d26b1df819b32a125fff2a79614684f4b772f9c6))


### Changed

* Enchant Conflict message ([5dfd81e](https://gitlab.com/bednic/json-api/commit/5dfd81e44ac4b3af6c33fbece5f5b35fc2dd3bed))
* remove final from DefaultErrorFactory ([d530347](https://gitlab.com/bednic/json-api/commit/d530347e285c36978527b12220e3c117226f1022))

src/Filter/URL.php
<?php
/**
 * Created by IntelliJ IDEA.
 * User: tomas
 * Date: 05.02.2019
 * Time: 13:19
 */

namespace JSONAPI\Filter;

use JSONAPI\LinkProvider;
use Psr\Http\Message\StreamInterface;

/**
 * Class URL
 *
 * @package JSONAPI
 */
class URL
{

    const EQUAL = "=";
    const NOT_EQUAL = "!";
    const GREATER_THEN = ">";
    const LOWER_THEN = "<";
    const LIKE = "~";
    const IN = "IN";

    const OFFSET = 'offset';
    const LIMIT = 'limit';

    private $includes = null;
    private $fields = null;
    private $sort = null;
    private $filter = null;
    private $pagination = [
        self::OFFSET => 0,
        self::LIMIT => 25
    ];

    /**
     * @var Endpoint
     */
    public $endpoint;

    /**
     * URL constructor.
     */
    public function __construct()
    {

        $this->endpoint = $this->parseEndpoint();

        if (isset($_GET['include'])) {
            $this->parseIncludes($_GET['include']);
        }

        if (isset($_GET['fields'])) {
            $this->parseFields($_GET['fields']);
        }
        if (isset($_GET['sort'])) {
            $this->parseSort($_GET['sort']);
        }
        if (isset($_GET['page'])) {
            $this->parsePage($_GET['page']);
        }

        if (isset($_GET['filter'])) {
            $this->parseFilter($_GET['filter']);
        }


    }

    /**
     * @return array
     */
    public function getIncludes(): array
    {
        return $this->includes;
    }

    /**
     * @param $resourceType
     * @return array
     */
    public function getFieldsFor($resourceType): array
    {
        return isset($this->fields[$resourceType]) ? $this->fields[$resourceType] : [];
    }

    /**
     * @return array
     */
    public function getSort(): array
    {
        return $this->sort;
    }

    /**
     * @return array
     */
    public function getPagination(): array
    {
        return $this->pagination;
    }

    /**
     * @return array|null
     */
    public function getFilter(): ?array
    {
        return $this->filter;
    }

    /**
     * @param string $query
     */
    private function parseIncludes(string $query)
    {
        $this->includes = [];
        $t = explode(",", $query);
        $dot2tree = function (&$arr, $path, $value, $separator = '.') {
            $keys = explode($separator, $path);
            foreach ($keys as $key) {
                $arr = &$arr[$key];
            }

            $arr = $value;
        };
        foreach ($t as $i) {
            $dot2tree($this->includes, $i, []);
        }
    }

    /**
     * @param array $query
     */
    private function parseFields(array $query)
    {
        $this->fields = [];
        foreach ($query as $type => $fields) {
            $this->fields[$type] = explode(',', $fields);
        }
    }

    /**
     * @param string $query
     */
    private function parseSort(string $query)
    {
        $this->sort = [];
        preg_match_all('/((?P<sort>-?)(?P<field>[a-zA-Z0-9]+))/', $query, $matches);
        foreach ($matches['field'] as $i => $field) {
            $this->sort[$field] = $matches['sort'][$i] ? "DESC" : "ASC";
        }
    }

    /**
     * @param array $pagination
     */
    private function parsePage(array $pagination)
    {
        if (isset($pagination[self::OFFSET])) {
            $this->pagination[self::OFFSET] = (int)$pagination[self::OFFSET];
        }
        if (isset($pagination[self::LIMIT])) {
            $this->pagination[self::LIMIT] = (int)$pagination[self::LIMIT];
        }
    }

    /**
     * It's custom idea, I can't guarantee that this will satisfy all needs
     *
     * @param array $filters
     */
    private function parseFilter(array $filters)
    {
        $this->filter = [];
        foreach ($filters as $field => $values) {
            if (is_array($values)) {
                foreach ($values as $value) {
                    preg_match('/^(?P<operand>!|>|<|~|)(?P<value>.+)/', $value, $matches);
                    $value = $this->guessDataType($matches["value"]);
                    if (is_array($value)) {
                        $operand = self::IN;
                    } elseif (in_array($matches["operand"], [self::GREATER_THEN, self::LOWER_THEN, self::NOT_EQUAL, self::LIKE])) {
                        $operand = $matches["operand"];
                    } else {
                        $operand = self::EQUAL;
                    }
                    $this->filter[$field][] = new Condition($value, $operand);
                }
            } else {
                preg_match('/^(?P<operand>!|>|<|~|)(?P<value>.+)/', $values, $matches);
                $value = $this->guessDataType($matches["value"]);
                if (is_array($value)) {
                    $operand = self::IN;
                } elseif (in_array($matches["operand"], [self::GREATER_THEN, self::LOWER_THEN, self::NOT_EQUAL, self::LIKE])) {
                    $operand = $matches["operand"];
                } else {
                    $operand = self::EQUAL;
                }
                $this->filter[$field] = new Condition($value, $operand);
            }

        }
    }

    /**
     * @param $value
     * @return mixed
     */
    private function guessDataType($value)
    {
        if (preg_match('/^\[([a-zA-Z0-9-_. ]+\,?)+\]$/', $value, $matches)) {
            preg_match_all('/([a-zA-Z0-9-_. ]+)\,?/', $matches[0], $values);
            $ret = $values[1];
            foreach ($ret as &$item) {
                $item = $this->guessDataType($item);
            }
            return $ret;
        } elseif ($ret = filter_var($value, FILTER_VALIDATE_INT)) {
            return $ret;
        } elseif ($ret = filter_var($value, FILTER_VALIDATE_FLOAT)) {
            return $ret;
        } elseif (($ret = filter_var($value, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE)) !== null) {
            return $ret;
        } elseif ($value === "null") {
            return null;
        } elseif (strlen($value) >= 4 && strtotime($value)) {
            try {
                return new \DateTime($value);
            } catch (\Exception $e) {
                return $value;
            }
        } else {
            return (string)$value;
        }
    }


    /**
     * @return Endpoint
     */
    private function parseEndpoint(): Endpoint
    {
        $baseUrl = LinkProvider::getUrl();
        $uri = "$_SERVER[REQUEST_SCHEME]://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
        $path = str_replace($baseUrl, '/', $uri);
        $pattern = '/^\/(?P<resource>[a-z-_]+)(\/(?P<id>[a-z0-9-_]+))?((\/relationships\/(?P<relationship>[a-z-_]+))|(\/(?P<related>[a-z-_]+)))?$/';
        preg_match($pattern, $path, $matches);
        return new Endpoint(
            $matches['resource'],
            isset($matches['id']) ? $matches['id'] : null,
            isset($matches['relationship']) ? $matches['relationship'] : null,
            isset($matches['related']) ? $matches['related'] : null
        );
    }
}
